### ISSUE TITLE
Feature: Allow case insensitive data lookup with IEQUAL etc

### ISSUE THREAD

**Issue by dereuromark** [propelorm/Propel2#1332]
We currently face some issues with case sensitivity and comparisons after switching from MySQL and CI by default to Postgres and CS by default.
Bugs have been reported regarding case sensitive data, e.g. for logging in with a specific email. Those fields shouldn't be case sensitive, and a hack quickfix would be to manually adjust all the queries everywhere in a rather dirty way.
It would be good if there was a more Propel way of fixing this globally.

It could be fixed via a setting globally, or for now also as a setting per "field".

### Solution a
Let's look into emails (which should be CI by definition) - we want to have CI here for sure:
```php
$query->filterByEmail($email);
```
This used to work with MySQL of course, as the collation was unicode ci.

One possible solution to be opt-in from the code perspective would be (adding new `Criteria::IEQUAL`):
```php
$query->filterByEmail($email, Criteria::IEQUAL);
```

It would then internally be "LOWER()" compared and return the expected result.
This would be consistent with `LIKE` and `ILIKE` in a way, but requires manual query adjusting.

### Solution b

Another one could be from the XML file perspective on generation (adding `caseInsensitive="true"`):
```xml
<column name="email" required="true" size="255" type="VARCHAR" caseInsensitive="true"/>
```

That means all query filters can stay as they are in code:
```php
$query->filterByEmail($email);
```
But will do the necessary adjustments of the query beneath the surface, so the using code doesn't need to care about it - which it shouldn't anyway if declared like this in the XML.

I prefer the latter as it would internally make all Criteria::EQUAL case insensitive the proper way without the business classes then have to differentiate. Otherwise you easily could miss a filter statement somewhere introducing bugs here. So fixing this on the lowest possible level in Propel2 makes sense to me. And depending on the database and collation it could then either add the internal query modifications or just omit it if not necessary (Mysql + CI).

Thanks for looking into this.

**Comment by marcj** [comment]
You complain about a missing feature and categorise it as a bug. There's something wrong: Either you didn't describe the actual bug (I don't see any bug yet) or you confuse bugs with missing features.

When you develop an application that can run on different database engines you should be aware of collation issues (ordering will change, comparing will change, grouping will change) right at the beginning. Common way to resolve this is to force a collation that is "nearly" equal in all databases. Means: the grouping of characters are as nearly together as possible and case sensitiveness is activated.

Depending on the scenario you need now to find a way to make CI possible per use case. Why per use case? Because it differs too much in the engines.

## Use UPPER() == UPPER().

Great thing is it works in all database okayish, but 

1. makes stuff verbose
2. **doesn't use indices** (unless you created upper indice, not everywhere supported)
3. breaks uniqueness expectations: if you use CS per default, you're allowed to add `foo`, `Foo`, `FOO` if you have a unique on it, but UPPER() makes it ambiguous which row to match. This issue is usually a nasty base for nasty bugs

You can either write custom query using 

```php
$query->where('UPPER(FieldA) = ?', strtoupper($v))
````

or let the Propel adapter handle it for simple conditions

```php
 //for all comparisons in that query
$query->setIngoreCase(true);

//for one particular condition
$query->filterByFieldA('bar');
$query->getCriterion(TableMAP::COL_FIELDA)->->setIngoreCase(true);
```

## caseInsensitive attribute

You might think a new attribute `<column caseInsensitive="true">` would be great, which activates
UPPER() comparison per default per column. However, you actually have the same issues as above except the verbose argument. So, this is not a way to go as it breaks way too many expectations about performance and indices.

## Use collation/type per column

This is the only "right" way of doing it, however you're now out of the scope of an ORM or a query builder. You dive now deep into each database specifics.

e.g. 

```xml
<column>
    <vendor type="mysql">
      <parameter name="Collate" value="utf8_unicode_ci"/>
    </vendor>
</column>
```

Is the only think you should use because it sustains the uniqueness, works fast with indices and you don't need to bother your PHP code to change the behavior.

One big drawback tho: You have to find for each database a way to say "this column is case insensitive".

1. SQLite it is `COLLATE UTF8CI` or NOCI
2. Postgres is using local locale so you need to specify `citext` as type
3. MySQL provides even more collations. You need to make sure you use the correct one for your use case, probably `utf8_unicode_ci` tho.

--------

So, as you see, there's no easy way that works in 99% across all databases we could use in Propel. It's too engine specific, too use case specific. Especially with collations you're very alone with each database and should optimize and think about that very early on.

**Comment by marcj** [comment]
> So fixing this on the lowest possible level in Propel2 makes sense to me.

Actually, it's vice-versa. You need to be aware of when exactly you need to do that. You already made that mistake in your application: Defining email as CS in the database and want to make it CI in the query. This breaks your SSOT and consistency in the database as the database allows you to insert "marc@" and "marC@", also this breaks the query once you have inconsistent data as it is ambiguous which row to fetch when you query `UPPER(email) == 'MARC@'`. So, my advice is to go through each query that needs CI and find a solution for each. Sometimes UPPER works, sometimes you need different collation because of performance and uniqueness. Making it magically workable in Propel is 1. not working as described above and 2. introduces wrong expectations (even you already fall into)

**Comment by dereuromark** [comment]
You are right. For us it is a bug, but here it should be classified as feature request. Changed the title.

The limitation itself can indeed be custom hacked on per app basis. But my point is: If the abstraction of ORM could handle this unified for a specific database type, that would be preferred over having to adjust each query.

>  You already made that mistake in your application

It wasn't a deliberate action, it was more a result from switching to MySQL to Postgres without having a test case for CS lookup, e.g. in login.

I agree with you that it has to be solved on a per field basis, but the "how" can make a huge difference.
I find it interesting that ILIKE exists as LIKE counterpart, but IEQUAL or anything would need custom solutions:

 > This is not in the SQL standard but is a PostgreSQL extension.

The extension could have also accounted for the above issue IMO.

Anyway, to sum it up:

- The actual code should ideally not have to become DB specific, but stay DB agnostic. Otherwise switching the DB underneath would force you to adjust all queries.
- Ideally it is solved as opt-in for those DBs where it is necessary, and ignored for those where it doesnt make a difference. Allowing a caseInsensitive option is also totally opt-in, and only affects the relevant DB. So this would be BC and allow others to use a more fine-grained solution as well. Totally right that we cannot solve it perfectly or 99% use case, but the 90% use case aim is usually a good fit for starters.

**Comment by marcj** [comment]
> I find it interesting that ILIKE exists as LIKE counterpart, but IEQUAL or anything would need custom solutions:

It's because PostgreSQL supports indices in free floating ILIKE patterns with GIN/GiST indices. That wouldn't be possible with `LOWER(bla) LIKE `. MySQL doesn't have something equal to ILIKE because it doesn't support powerful indices like GIN/GIST out-of-the-box and you'd use fulltext columns with a totally different comparator (`MATCH ... AGAINST`).

> It wasn't a deliberate action, it was more a result from switching to MySQL to Postgres without having a test case for CS lookup, e.g. in login.

Yeah, I know. :) That's why I meant case sensitiveness should have a big awareness right at the start of development, especially when you want to support multiple databases. This is not an easy feature to implement, especially not if you have very big databases and you **need** to have fast queries that use indices.

-----

In my opinion it's out of context of an ORM/Query builder. We provide already anything to make your use case workable. Your wish is to make it even easier by having a column attribute. However, in my opinion it introduces the sham that it magically works as described ("What can be so hard to make a query case insensitive, hu?") This fallacy leads to many new issues in the application you'll see as soon as the application grows. 

Making a UPPER() == UPPER() is a very explicit decision with all its drawbacks. Hiding this behind `caseInsensitive` appears to me wrong. It should remain explicit also in your php source code.

**Comment by dereuromark** [comment]
According to https://www.postgresql.org/docs/9.1/static/citext.html it sure seems like using this `citext` here on the DB level is indeed the most viable solution for those varchar fields that need CI consistently - the article also states why other solutions are not so good, like danger of inconsistently using the verbose declaration, opening up for possible data inconsistencies.

Does Propel offer any ways of declaring this in the XML files?

**Comment by marcj** [comment]
it's basically `<column sqlType="citext">`. This binds however the schema directly to PostgreSQL.

**Comment by dereuromark** [comment]
That's what I am a bit afraid of. As vendor/framework we must stay DB agnostic wherever possible.
So if this is not something Propel can or want to abstract away, we need to extend and customize on our ends, e.g. with the caseInsensitive="true" XML flag, which then will translate into citext type for Postgres and CI collation for MySQL.
We will have to discuss the options here - and we are always happy about further tips, feedback and improvements from your end.
Thanks a lot!

**Comment by oliwier-spryker** [comment]
There is a problem with `<column sqlType="citext">`, Postgres and migrations.

I've made a bug report about, [please take a look here](https://github.com/propelorm/Propel2/issues/1336).

**Comment by dereuromark** [comment]
I think this got finally then solved through https://github.com/propelorm/Propel2/issues/1336
